content/learning-paths/smartphones-and-mobile/ray_tracing/rt06_reflections.md

```diff
@@ -0,0 +1,439 @@
+
+When we think of reflections, we usually think of mirror like reflections, however non mirror objects also reflect part of their incident light producing glossy or rough reflections. In a PBR render we use the material roughness to decide the reflected light dispersion. A roughness of 0 will indicate that the material is a mirror, so we will need a single ray to solve it. On the other hand, objects with a higher roughness will reflect the incident ray in multiple directions, producing glossy reflections. A render will need to launch multiple rays to solve the color of glossy reflections.
+
+``` glsl
+vec4 main()
+{
+    o_color = vec4(0);
+    // Retrieve some data from the G-buffer
+    vec2 g_buffer_uv = vec2(gl_FragCoord.xy) / current_resolution;
+
+    ivec2 texture_size = textureSize(depth_sampler, 0);
+    ivec2 depth_coord_texel = ivec2(g_buffer_uv * texture_size);
+    float depth = get_g_buffer_depth(depth_coord_texel);
+    if (depth <= 0)
+        return;
+
+    const vec3 pos = get_g_buffer_world_pos(g_buffer_uv);
+    const vec4 material_properties = get_g_buffer_material_properties(g_buffer_uv);
+    const vec3 normal = get_g_buffer_normal(g_buffer_uv);
+    const bool is_reflective = get_material_is_reflective(material_properties);
+    const float roughness = get_material_roughness(material_properties);
+
+    // Reflection ray
+    const int MAX_REFLECTIONS_SAMPLES = 1;
+    bool is_valid_ray = is_reflective;
+    if (is_valid_ray)
+    {
+        const bool is_mirror_reflections = MAX_REFLECTIONS_SAMPLES <= 1 || roughness < 0.001;
+
+        const vec3 incident_dir = normalize(pos - camera_position);
+        vec4 reflection_color = vec4(0);
+        if (is_mirror_reflections)
+        {
+            launch_reflection_ray(pos, incident_dir, normal, reflection_color);
+        }
+        else
+        {
+            for (uint i = 0; i < MAX_REFLECTIONS_SAMPLES; i++)
+            {
+                vec2 epsilon = get_blue_noise(i);
+                vec4 ggx_sampled_half = get_ggx_sampled_half(roughness, epsilon, normal);
+
+                launch_reflection_ray(pos, incident_dir, ggx_sampled_half, reflection_color);
+            }
+            reflection_color /= float(MAX_REFLECTIONS_SAMPLES);
+        }
+        o_color += reflection_color;
+    }
+}
+```
+
+Launching multiple rays can be expensive, so we recommend limiting the number of rays in your shaders. At the same time, glossy reflections will launch rays with more divergence, being more expensive than pure mirror reflections. It is more efficient to traverse the acceleration structure if all rays in the same warp follow a similar path, so try to minimize ray divergence. If we want glossy reflections, we might need to add a denoising pass to deal with this noise. This will further increase the cost of glossy reflections. Please considered the performance impact before adding glossy reflections
```

Comment:
last sentence has a couple of typos, should be "Please consider the performance impact before adding glossy reflections."